### PR TITLE
Improve address handling abstraction

### PR DESCRIPTION
**Pull request by pitrou**
This removes some of the roadblocks for using non-IP based protocols for local communication (notably, the notion of a "port" does not exist with `inproc` communications).

Based on PR #888.

### CODE REVIEW

File: distributed/bokeh/background/server_lifecycle.py [hunk 1]
```diff
@@ -124,11 +126,6 @@ def task_events(interval, deque, times, index, rectangles, workers, last_seen):
         pass  # don't log CommClosedErrors
     except Exception as e:
         logger.exception(e)
-    finally:
```
Comment (pitrou) [inline distributed/bokeh/background/server_lifecycle.py]: Note this block was a no-op, as `sys.exit` always raises SystemExit.

**Comment by mrocklin** [comment]
My first impression is that things seem clean, but also that the class hierarchy around comms is growing significantly more complex.  I think that now would be a good time to build real documentation around it so that others can easily understand this code after reading sufficient prose.

**Comment by pitrou** [comment]
Thank you, good point about the documentation. i'll add some to make the architecture more understandable.

**Comment by pitrou** [comment]
I think I've improved the documentation now. I'm not sure how informative that is, so please tell me if you think something is missing.

**Comment by mrocklin** [comment]
cc @jcrist , who has expressed an interest to learn more about dask/distributed communication

File: docs/source/communications.rst [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+Communications
+==============
+
+Workers, the Scheduler, and Clients communicate by sending each other
+Python objects (such as :ref:`protocol` messages or user data).
+The communication layer handlers appropriate encoding and shipping
+of those Python objects between the distributed endpoints.  The
+communication layer is able to select between different transport
+implementations, depending on user choise or (possibly) internal
+optimizations.
```
Comment (mrocklin) [inline docs/source/communications.rst]: handlers -> handles

choise -> choice

**Comment by pitrou** [comment]
@jcrist, do you want to take a look soon or should I merge?

**Comment by mrocklin** [comment]
I think that @jcrist is busy today.  I recommend merging.

**Pull request merged**